packages/labs/ssr/src/lib/module-loader.ts

```diff
@@ -0,0 +1,279 @@
+
+const builtIns = new Set(builtinModules);
+
+const specifierMatches = (specifier: string, match: string) =>
+  specifier === match || specifier.startsWith(match + '/');
+
+export interface ModuleRecord {
+  path: string;
+  module?: vm.Module;
+  imports: Array<string>;
+  evaluated: Promise<vm.Module>;
+}
+
+interface ImportResult {
+  path: string;
+  module: vm.Module;
+}
+
+/**
+ * A JavaScript module loader that utilizes the Node `vm` module
+ * (https://nodejs.org/api/vm.html).
+ *
+ * Most of the hooks implement fairly standard web-compatible module loading:
+ *  - An import specifier resolver that uses Node module resoution
+ *  - A linker that loads dependencies from the local filesystem
+ *  - A module cache keyed by resolved URL
+ *  - import.meta.url support
+ *  - Dynamic import() that functions the same as static imports
+ *
+ * There are some behaviors specific to lit-html. Mainly that imports of certain
+ * directives are redirected to Node/SSR compatible implementations.
+ */
+export class ModuleLoader {
+  private static _nextVmContextId = 0;
+
+  // TODO (justinfagnani): document the need for this
+  private readonly _vmContextId = ModuleLoader._nextVmContextId++;
+
+  private readonly _context: vm.Context;
+
+  /**
+   * TODO (justinfagnani): This is a temporary stand-in for a real graph API.
+   * We want to be able to invalidate a module and the transitive closure
+   * of its importers so that we can update the graph.
+   *
+   * The keys of the map are useful for enumering static imported modules
+   * after an entrypoint is loaded.
+   */
+  readonly cache = new Map<string, ModuleRecord>();
+
+  // TODO (justinfagnani): Allow passing a filesystem object to allow network
+  // sources, in-memory for tests, etc.
+  constructor(contextGlobal: vm.Context) {
+    this._context = vm.createContext(contextGlobal);
+  }
+
+  /**
+   * Imports a module given by `path` into a new VM context with `contextGlobal` as the
+   * global object.
+   *
+   * @param specifier
+   * @param referrer
+   * @param contextGlobal The object that will become the global, via vm.createContext
+   */
+  async importModule(
+    specifier: string,
+    referrer: string
+  ): Promise<ImportResult> {
+    const result = await this._loadModule(specifier, referrer);
+    const {module} = result;
+    if (module.status === 'unlinked') {
+      await result.module.link(this._linker);
+    }
+    if (module.status !== 'evaluated') {
+      await result.module.evaluate();
+    }
+    return result;
+  }
+
+  /**
+   * Performs the actual loading of module source from disk, creates the
+   * Module instance, and maintains the module cache.
+   *
+   * Used directly by `importModule` and by the linker and dynamic import()
+   * support function.
+   */
+  private async _loadModule(
+    specifier: string,
+    referrer: string
+  ): Promise<ImportResult> {
+    if (/:\d+$/.test(referrer)) {
```

Comment:
Isn't `:0` a valid filename?

Reply:
I can remove this. It was useful for flushing out bugs...

Reply:
I guess I was wondering about the prefixing scheme overall, although I suppose it doesn't actually require that there are never filenames that match the prefix pattern.